conftest.py

```diff
@@ -51,3 +55,47 @@ def pytest_itemcollected(item):
 
     if isinstance(item, pytest.DoctestItem):
         item.add_marker(doctest_marker)
+
+
+def pytest_addoption(parser: pytest.Parser) -> None:
+    """Hook to register custom CLI options and config values"""
+    parser.addoption(
+        "--strict-warnings",
+        action="store_true",
+        default=False,
+        help="Turn most warnings into errors",
+    )
+
+    parser.addini(
+        "filterwarnings_when_strict",
+        "Filters to apply after `-Werror` when --strict-warnings is active",
+        type="linelist",
+        default=[],
+    )
+
+
+def pytest_collection_modifyitems(
+    session: pytest.Session, config: pytest.Config, items: Iterable[pytest.Item]
+):
+    if not config.getoption("--strict-warnings"):
+        return
+
+    warning_filters = [
+        r"error",
+        *_config_get_strlist(config, "filterwarnings"),
+        *_config_get_strlist(config, "filterwarnings_when_strict"),
+    ]
+    warning_marks = [pytest.mark.filterwarnings(f) for f in warning_filters]
+
+    for item in items:
+        # reverse and then prepend means essentially `marks[0:0] = warning_marks`
```

Comment:
What is `marks` here?

Reply:
Also just a little lost as to what the objective is (or rather why reversing assists in that)

Reply:
Ah, ok, I see.  I guess you want the users CLI options to override ours here.  But not sure I understand the reversal.

Reply:
To preserve the order, as the comment says:

```py
before = ["last"]
to_prepend = ["first", "middle"]

correct = before.copy()
for item in reversed(to_prepend):
    correct.insert(0, item)
assert correct = ["first", "middle", "last"]

wrong = before.copy()
for item in to_prepend:
    wrong.insert(0, item)
assert wrong == ["middle", "first", "last"]
```

Reply:
Ok, I guess I see now.  You could have overrides within the list being prepended that you want applied in order.